Anthonyw298

anthonyw298 opened the pull request:
Changed layout of Base Code where I took away fans and kept the same files as main branch. Only changed the pump variables by adding a structure type

anthonyw298 committed:
Added a fan cycle code where fan A is on 1s theb B is on 1s then both are one 1s then both are off 3s before cycling thru again

Also reoutlined the code by using struct to make it more nice to view

anthonyw298 committed:
Added Code for Pump.ino Unit testing utilizing the same structure code as fans.ino unit testing code

THis code will initialize the pump to move with the testing of the pump speed and being able to adjust it

anthonyw298 committed:
Revised Base.ino Code

Here i revised the code where I labeled various struct types that initializes the related variables&values within the function that corresponds to the section of the water pump system. I then installed the various functions to run the smart system after these structures. FInally I call these functions and run it for a loop. This code includes the fan cycle and water pump cycle using PWM.

anthonyw298 committed:
Unit Testing for pump.ino

Added Code for Pump.ino Unit testing utilizing the same structure code as fans.ino unit testing code. This code will initialize the pump to move with the testing of the pump speed and being able to adjust it

anthonyw298 committed:
Variable Grouping Of Pump

Took away the fan and other outside parts of code and just focused on varuable grouping the pump variables into a structure type and then kept the original code and functions for the rest

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -1,112 +1,142 @@
-      PumpPWM = 85; // Set the pump speed to medium-low
-      previousTime = currentTime; // Update the time of the last water level check
+// Initialize the fans
+void beginFans(Fans &fans) {
```
dont create this function

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -1,112 +1,142 @@
+}
+
+// Update fan control based on time
+void updateFans(Fans &fans, unsigned long currentTime) {
```
dont create this function

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -1,112 +1,142 @@
-  waterLevelB = digitalRead(WaterLevelB);
-  waterLevelC = digitalRead(WaterLevelC);
+// Read water levels from sensors
+void readWaterLevels(WaterLevelSensors &sensors, int &levelA, int &levelB, int &levelC) {
```
dont create this function

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -1,112 +1,142 @@
+};
+
+// Initialize the water level sensors
+void beginWaterLevelSensors(WaterLevelSensors &sensors) {
```
dont create this function

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -1,112 +1,142 @@
+};
+
+// Define a structure for water pump control
+struct WaterPumpController {
```
this is how I want a structure to look

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -1,112 +1,142 @@
+    const unsigned long pumpTimeOff;
+    const unsigned long pumpTimeOn;
+
+    WaterPumpController(int p)
```
dont initialize inside the struct however

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -1,112 +1,142 @@
+    unsigned long previousTime;
+    int state;
+
+    Fans(int aPin, int bPin) : pinA(aPin), pinB(bPin), previousTime(0), state(0) {}
```
dont initialize inside the structure

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -1,112 +1,142 @@
+    int pinB;
+    int pinC;
+
+    WaterLevelSensors(int a, int b, int c) : pinA(a), pinB(b), pinC(c) {}
```
dont initialize inside the structure, initialization occurs in void setup

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -1,112 +1,142 @@
-const int WaterPump = 11; // Pin for the water pump
-int PumpPWM; // Variable to control the speed of the water pump
-
-void setup() 
```
this is how I want void setup to be, don't initialize inside structures

anthonyw298 committed:
Variable Grouping Of Pump

Took away the fan and other outside parts of code and just focused on varuable grouping the pump variables into a structure type and then kept the original code and functions for the rest

anthonyw298 committed:
Test

anthonyw298 committed:
Test

anthonyw298 committed:
Base.ino Variable Grouping

Here is the actual revised code. based off the previous comments you had, I added a structure that grouped all the variables for the pump and kept the rest of the code the same.

anthonyw298 committed:
Base.ino Variable Grouping

Here is the actual revised code. based off the previous comments you had, I added a structure that grouped all the variables for the pump and kept the rest of the code the same.

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -1,5 +1,5 @@
 #include <Arduino.h>
-
+//Actual Revised Code
 // Define the pins for the water level sensors
```
Correct me if I am wrong but I think we should group all the water level sensors together in a struct. So both the pin declarations and the variables that stores the water levels can be in the same struct.

david1534 replied:
We should also rename any pin declarations from WaterLevelA to WaterPinA format so it is easier to tell between variables and pins

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -11,15 +11,22 @@ const int FanA = 9;
 const int FanB = 10;
 int FanPWM; // Variable to control the speed of the fans
```
Same thing with the fans, group into struct and rename

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -11,15 +11,22 @@ const int FanA = 9;
+// Create an instance of the WaterPumpFn struct
+WaterPumpFn waterPump;
+
 const int WaterPump = 11; // Pin for the water pump
```
Lets throw this in the struct

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -106,7 +113,6 @@ void loop()
   unsigned long currentTime, elapsedTime, currentPumpTime, elapsedPumpTime; // Variables to store the current time and elapsed time
```
Can these variables all be grouped with their respective struct? maybe we can simply create an instance in the loop if needed

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -106,7 +113,6 @@ void loop()
   unsigned long currentTime, elapsedTime, currentPumpTime, elapsedPumpTime; // Variables to store the current time and elapsed time
 
   checkWaterLevel(waterLevelA, waterLevelB, waterLevelC); // Check the current water level
```
The point of the struct is to have it passed into functions all at once. For example, checkWaterLevel(WaterLevelStruct);

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -106,7 +113,6 @@ void loop()
 
   checkWaterLevel(waterLevelA, waterLevelB, waterLevelC); // Check the current water level
-  checkPumpStatus(currentTime, elapsedTime, currentPumpTime, elapsedPumpTime, 
+  checkPumpStatus(waterPump, currentTime, elapsedTime, currentPumpTime, elapsedPumpTime, 
   waterLevelA, waterLevelB, waterLevelC); // Check and update the pump status
```
Same thing here

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -44,59 +51,59 @@ void checkWaterLevel(int &waterLevelA, int &waterLevelB, int &waterLevelC)
 
 // Function to check and update the status of the pump
-void checkPumpStatus(unsigned long &currentTime, unsigned long &elapsedTime, 
+void checkPumpStatus(WaterPumpFn &waterPump, unsigned long &currentTime, unsigned long &elapsedTime, 
```
See comment at the bottom, but the point of a struct is to group similar variables or pin declarations together

david1534 reviewed (changes requested):
Let me know if these changes are feasible to implement. I'm not 100% sure if it is so I wanted you to explore the proposed changes and validate it with the electrical simulation before you update this pull request. Do not delete this current request, I will resolve the comments in this review when the necessary changes are addressed or corrected.

anthonyw298 committed:
Test

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -2,46 +2,63 @@
-const unsigned long pumpTimeOn = 5000; // Time in milliseconds for the pump to stay on
-const int WaterPump = 11; // Pin for the water pump
-int PumpPWM; // Variable to control the speed of the water pump
+// Define the structs
+
+struct WaterLevelSensors {
+  const int WaterPinA = 4;
+  const int WaterPinB = 5;
+  const int WaterPinC = 6;
+  const int WaterPinD = 7;
+  int waterLevelA;
+  int waterLevelB;
+  int waterLevelC;
+  int waterLevelD;
+};
+
+struct Fans {
+  const int FanA = 9;
+  const int FanB = 10;
+  int FanPWM;
+};
+
+struct WaterPumpFn {
+  unsigned long previousTime = 0;
+  unsigned long previousPumpTime = 0;
+  const unsigned long stabilityTime = 1000;
+  bool pumpStatus = false;
+  const unsigned long pumpTimeOff = 1000;
+  const unsigned long pumpTimeOn = 5000;
+  const int WaterPump = 11;
+  int PumpPWM;
+  unsigned long currentTime, elapsedTime, currentPumpTime, elapsedPumpTime;
+};
```
Is it possible to have these structures defined in its own header file. One specifically for structures?

anthonyw298 replied:
not sure what you mean by this

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -2,46 +2,63 @@
+  Serial.begin(9600);
+
+  // Set the mode of the pins using the struct instances
+  pinMode(waterLevelSensors.WaterPinA, INPUT);
+  pinMode(waterLevelSensors.WaterPinB, INPUT);
+  pinMode(waterLevelSensors.WaterPinC, INPUT);
+  pinMode(waterLevelSensors.WaterPinD, OUTPUT);
+  pinMode(fans.FanA, OUTPUT);
+  pinMode(fans.FanB, OUTPUT);
+  pinMode(waterPumpFn.WaterPump, OUTPUT);
```
Good use of structure

david1534 commented on Simulation Code/Base/Base.ino:
```diff
@@ -2,46 +2,63 @@
 {
-  int waterLevelA, waterLevelB, waterLevelC, waterLevelD; // Variables to store the current water level
-  unsigned long currentTime, elapsedTime, currentPumpTime, elapsedPumpTime; // Variables to store the current time and elapsed time
+  waterPumpFn.currentTime = millis();
```
I'm assuming this has to be declared within the loop?

anthonyw298 replied:
yeah since its going to be constantly running

david1534 commented on Simulation Code/Base/README.txt:
Add a readme file for descriptions of each of the files within this folder

david1534 reviewed (changes requested):
This is good and basically ready for merging. See comments @anthonyw298

anthonyw298 committed:
Updated Code With Header Files

Used struct in the .h files to declare all variables within the pump and waterlevel sensor code. The .cpp has the code to run the function and base.ino declares input/output, loops of function, and creates instances of the structs. Verified this code on ArduinoIDE and tested on SimulIDE and both are good.

anthonyw298 committed:
Updated Code With Header Files

Used struct in the .h files to declare all variables within the pump and waterlevel sensor code. The .cpp has the code to run the function and base.ino declares input/output, loops of function, and creates instances of the structs. Verified this code on ArduinoIDE and tested on SimulIDE and both are good.
